### PR TITLE
#1750 - Using 'd0' to delete to start of the current logical line in a multiline buffer

### LINKED ISSUE #1750: [vi-mode]: Using 'd0' should only delete up to the beginning of the current logical line in a multiline buffer
## Description of the new feature/enhancement

Using <kbd>d</kbd><kbd>0</kbd> currently deletes the entire selection from the start of the buffer to the cursor position. In a multiline buffer, this should rather delete to the beginning of the current logical line only.

### PR DESCRIPTION
**Pull request by springcomp**
# PR Summary

Fixed #1750.

This PR changes the behaviour of the `BackwardDeleteLine` method to delete characters up to the start of the current logical line, instead of up to the start of the buffer. This makes it more consistent with other changes in `Vi` mode related to multiline buffers.

I’m concerned this may have an impact on `Windows` mode but could not see regression test failures. So I’m not sure this scenario is covered. In any case, I have created a new method `BackwardDeleteInput` for this purpose and would be happy to update documentation accordingly.

## PR Checklist

- [x] PR has a meaningful title
    - Use the present tense and imperative mood when describing your changes
- [x] Summarized changes
- [x] Make sure you've added one or more new tests
- [x] Make sure you've tested these changes in terminals that PowerShell is commonly used in (i.e. conhost.exe, Windows Terminal, Visual Studio Code Integrated Terminal, etc.)
- **User-facing changes**
    - [ ] Not expecting documentation change.
    - **OR**
    - [x] Documentation needed at [PowerShell-Docs](https://github.com/MicrosoftDocs/PowerShell-Docs)
        - [x] Doc Issue filed: [PowerShell-Docs#7146](https://github.com/MicrosoftDocs/PowerShell-Docs/pull/7146)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/PowerShell/PSReadLine/pull/2002)

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@springcomp Sorry for being sluggish. The changes make perfect sense, but it raises more inconsistency to the function names that are responsible for deleting texts from the input. For examples, after the change, `BackwardDeleteLine` deletes a logical line, but
1. `ForwardDeleteLine` deletes till the end of input;
2. `BackwardKillLine` deletes from the start of the input;

I know there are already inconsistency existing, such as `DeleteLine` deleting the current logical line while `KillLine` somehow works like `ForwardDeleteLine` and deletes till the end of buffer ... Fixing those existing inconsistency is out of scope for this PR.

For this PR, can we maybe bring consistency to `ForwardDeleteLine` and `BackwardKillLine`? What I mean is that:
1. Have `ForwardDeleteInput` for deleting till the end of input, and make `ForwardDeleteLine` delete to the end of current logical line.
2. Have `BackwardKillInput` for kill texts from the start of input, and make `BackwardKillLine` kill texts from the start of the current logical line.

What do you think?

File: PSReadLine/BasicEditing.cs [hunk 1]
```diff
@@ -115,17 +115,33 @@ public static void ForwardDeleteLine(ConsoleKeyInfo? key = null, object arg = nu
         }
 
         /// <summary>
-        /// Like BackwardKillLine - deletes text from the point to the start of the line,
+        /// Like BackwardKillLine - deletes text from the point to the start of the buffer,
+        /// but does not put the deleted text in the kill ring.
+        public static void BackwardDeleteBuffer(ConsoleKeyInfo? key = null, object arg = null)
```
Comment (daxian-dbw) [inline PSReadLine/BasicEditing.cs]: Maybe name it `BackwardDeleteInput`? `Buffer` is more of an implementation detail which may be confusing to the user.

Comment (springcomp) [reply PSReadLine/BasicEditing.cs]: Agreed.

File: PSReadLine/KeyBindings.cs [hunk 2]
```diff
@@ -388,10 +388,11 @@ public static KeyHandlerGroup GetDisplayGrouping(string function)
             case nameof(AcceptAndGetNext):
             case nameof(AcceptLine):
             case nameof(AddLine):
+            case nameof(BackwardDeleteInput):
             case nameof(BackwardDeleteChar):
             case nameof(BackwardDeleteLine):
             case nameof(BackwardDeleteWord):
-            case nameof(BackwardKillLine):
+            case nameof(BackwardKillInput):
```
Comment (daxian-dbw) [inline PSReadLine/KeyBindings.cs:395]:
```suggestion
            case nameof(BackwardKillInput):
            case nameof(BackwardKillLine):
```

File: PSReadLine/KeyBindings.cs [hunk 3]
```diff
@@ -408,7 +409,7 @@ public static KeyHandlerGroup GetDisplayGrouping(string function)
             case nameof(DeletePreviousLines):
             case nameof(DeleteToEnd):
             case nameof(DeleteWord):
-            case nameof(ForwardDeleteLine):
+            case nameof(ForwardDeleteInput):
```
Comment (daxian-dbw) [inline PSReadLine/KeyBindings.cs:413]:
```suggestion
            case nameof(ForwardDeleteInput):
            case nameof(ForwardDeleteLine):
```

**Review by daxian-dbw — COMMENTED** [review]
Mostly look good to me. Some `briefDescription` are not right in `KeyBindings`.

**Review by daxian-dbw — APPROVED** [review]

**Pull request merged**